src/blockchain_db/lmdb/db_lmdb.h

```diff
@@ -45,126 +46,215 @@ typedef struct txindex {
+  tx_indices,
+  tx_outputs,
+
+  spent_keys,
+
+  txpool_meta,
+  txpool_blob,
+
+  alt_blocks,
+
+  hf_versions,
+
+  properties
+};
+
+class mdb_txn_cursors
 {
-  bool m_rf_txn;
-  bool m_rf_blocks;
-  bool m_rf_block_heights;
-  bool m_rf_block_info;
-  bool m_rf_output_txs;
-  bool m_rf_output_amounts;
-  bool m_rf_txs;
-  bool m_rf_txs_pruned;
-  bool m_rf_txs_prunable;
-  bool m_rf_txs_prunable_hash;
-  bool m_rf_txs_prunable_tip;
-  bool m_rf_tx_indices;
-  bool m_rf_tx_outputs;
-  bool m_rf_spent_keys;
-  bool m_rf_txpool_meta;
-  bool m_rf_txpool_blob;
-  bool m_rf_alt_blocks;
-  bool m_rf_hf_versions;
-  bool m_rf_properties;
-} mdb_rflags;
-
-typedef struct mdb_threadinfo
+  public:
+    mdb_txn_cursors() noexcept :
+      m_cursors{{nullptr}}
+    {}
+
+    ~mdb_txn_cursors() noexcept
+    {
+      for (auto cursor: m_cursors) {
+        if (cursor != nullptr) {
+          mdb_cursor_close(cursor);
+        }
+      }
+    }
+
+    MDB_cursor*& get(source cursor)
+    {
+      return m_cursors.at(static_cast<uint8_t>(cursor));
+    }
+  private:
+    std::array<MDB_cursor*, 18> m_cursors;
+};
+
+class mdb_databases
 {
-  MDB_txn *m_ti_rtxn;	// per-thread read txn
-  mdb_txn_cursors m_ti_rcursors;	// per-thread read cursors
-  mdb_rflags m_ti_rflags;	// per-thread read state
+    public:
+      mdb_databases() :
+        m_databases{{}}
+      {}
+
+      const static std::string& name(source database);
+
+      const MDB_dbi& get(source database) const
+      {
+        return m_databases.at(static_cast<uint8_t>(database));
+      }
+
+      MDB_dbi& get(source database)
+      {
+        return m_databases.at(static_cast<uint8_t>(database));
+      }
+
+      MDB_stat stat(MDB_txn* txn, source database) const;
+      void drop(MDB_txn* txn, source database, bool close = false);
+    private:
+      std::array<MDB_dbi, 18> m_databases;
+};
 
-  ~mdb_threadinfo();
-} mdb_threadinfo;
 
-struct mdb_txn_safe
-{
-  mdb_txn_safe(const bool check=true);
-  ~mdb_txn_safe();
 
-  void commit(std::string message = "");
+struct mdb_threadinfo
+{
+  mdb_threadinfo() :
+    m_ti_rtxn{nullptr},
+    m_rf_txn{false},
+    m_handles{0}
+  {}
 
-  // This should only be needed for batch transaction which must be ensured to
-  // be aborted before mdb_env_close, not after. So we can't rely on
-  // BlockchainLMDB destructor to call mdb_txn_safe destructor, as that's too late
-  // to properly abort, since mdb_env_close would have been called earlier.
-  void abort();
-  void uncheck();
+  ~mdb_threadinfo();
 
-  operator MDB_txn*()
-  {
-    return m_txn;
-  }
+  MDB_txn* transaction(MDB_env* env, bool *started = nullptr);
+  void reset();
 
-  operator MDB_txn**()
-  {
-    return &m_txn;
-  }
+  MDB_txn* acquire(MDB_env* env);
+  void release();
 
-  uint64_t num_active_tx() const;
+  MDB_txn *m_ti_rtxn;	// per-thread read txn
+  mdb_txn_cursors m_ti_rcursors;	// per-thread read cursors
+  bool m_rf_txn;
+  std::bitset<18> m_valid_cursor;
```

Comment:
No. There are basically three states when it comes to read cursors:

1. Not initialized (a nullptr)
2. Created using the current transaction (not a nullptr)
3. Created using an aborted transaction (not a nullptr)

State 3 is useful because it allows us to _re-use_ a cursor that was previously allocated for us by LMDB. However, since 2 and 3 both have a read cursor that is not a nullptr we need something to distinguish between these two states.

From [the LMDB documentation for mdb_cursor_renew](http://www.lmdb.tech/doc/group__mdb.html#gac8b57befb68793070c85ea813df481af):

> Renew a cursor handle.
> 
> A cursor is associated with a specific transaction and database. Cursors that are only used in read-only transactions may be re-used, to avoid unnecessary malloc/free overhead. The cursor may be associated with a new read-only transaction, and referencing the same database handle as it was created with. This may be done whether the previous transaction is live or dead.

This feature was already used by the `blockchain_db`, however, since I needed to change the `mdb_txn_cursors` struct to use a real array to fix the UB, I also changed the read flags to a real array (in this case an `std::bitset` because it's more-or-less an array of bools).

Reply:
I know how the cursors work, but I hadn't read the relevant code that used this, so made the cardinal sin of commenting too early.